### ISSUE TITLE
Receive OK, but sending does not work with espurna

### ISSUE THREAD

**Issue by franki29** [Portisch/RF-Bridge-EFM8BB1#2]
Hello, i could manage to load the prebuild hex image to my sonoff rfbridge.
Now i tried to learn a code from my remote and it is working.
But i could not send any learnd command. 

I am using pilight with my raspberry pi for sniffering and my bridge does not send out any code.

Any help what i have to do to get it work?

**Comment by Portisch** [comment]
I don't know the status of the espurna firmware if every commands are supported.

**Comment by franki29** [comment]
Hi, can you tell us how you are using the code? I tested it also with Tasmota Software, but same problem.
Looks like the original command does not work? What are you using to send the code?
Thanks for helping

**Comment by Portisch** [comment]
I do all the tests with a COM terminal program like HTerm.

**Comment by franki29** [comment]
To what port do you connect and with what baud rate?

**Comment by Portisch** [comment]
To the RX/TX header next to the EFM8BB1. And put the switch to the OFF position. This will cancel the RS232 connection to the ESP ISP.
Baud: 19200, 8N1

**Comment by franki29** [comment]
Hi , i tried with your way with HTerm, but still nothing was send. I tried also the sample code on your page.
My code would be AAA52724014003D405151155 and it should be OK.
The only code that make some reaction is the learning command.
And of course receiving is no problem.
Anny suggestion what i am doing wrong ?

**Comment by Portisch** [comment]
AA Sync
A5 Comand
2724 T-Sync
0140 T-Low
03D4 T-high
051511 Data
55 Sync end

Looks alright. I will test it later this week.

**Comment by ciotlosm** [comment]
I've also flashed the provided hex file and I'm able to learn commands but they don't work when sent back. I assume there is a problem with sending.

**Comment by franki29** [comment]
Hi, the code you put as "original" does not work.

**Comment by Portisch** [comment]
I checked the transmitting and there is a problem of handling the SYN115 chip.
Does somebody have any information (source code) how to handle the chip correct?

The SYN115 chip will transfer the data but the sync isn't correct. So the receiver will not understand the signal.

**Comment by AlbertWeterings** [comment]
I hope having a look [here](http://www.airspayce.com/mikem/arduino/RadioHead/) will help.

**Comment by Portisch** [comment]
No, they just attach dummy data in front to enable the chip. Also they don't have a pause of 10ms in the transfer.

**Comment by AlbertWeterings** [comment]
I will have to wait till I get a new Bridge with original firmware than I can log what is happening on the data line. I hope I will also be able to read the original firmware from the chip. I found that with my Wellon VP998 I can read the data from the EFM8 chip as long as the code protection bit is not set.

**Comment by Portisch** [comment]
Chip protection isn't set. Also I asked them for the original HEX file - no support.

**Comment by Portisch** [comment]
I just checked it today again. It looks like working with the last commit.

I have done a test by sending the data AAA5382C01CC057845D50C55 4 times, with a 200ms delay to the EFM8 chip:
![scope_0](https://user-images.githubusercontent.com/3938570/37195508-6af4c262-2373-11e8-92e1-828aac84a0f5.png)
Channel 1 is the data line on the SYN115. Channel 2 is the data line on the receiver.
As you can see one difference to remotes is the pause between the data packets.
A normal remote will send these 4 packets without a delay of 200ms between each data.

![scope_1](https://user-images.githubusercontent.com/3938570/37195583-c1732674-2373-11e8-9902-efde188f69ce.png)
This is a detail of one packet. As you can see the sync and data is received by the EFM8 itself right.

If somebody do have an oscilloscope and original firmware a record of transmitting would be helpfull.

Here the picture where I attached the oscilloscope:
![sonoff](https://user-images.githubusercontent.com/3938570/37195752-66748c1c-2374-11e8-9f6b-533d4ae2d345.png)

**Comment by AlbertWeterings** [comment]
@Portisch I have a oscilloscope only have to wait on a new bridge with original firmware. I ordered a few but it seems the (delivery)pigeon have difficulties finding the Netherlands. So when they arrive I hope to have the original firmware available and can do some measurements.

**Comment by franki29** [comment]
Hi , thanks to keep on going. Maybe this link helps : https://drive.google.com/open?id=0BwEYW5Q6bg_ZTDhKQXphN0ZxdEU. Original web side is http://www.rflink.nl/blog2/development. It looks like they support two SYN115 devices. http://www.rflink.nl/blog2/wiring : Cheap Alternative transmitter 2. Hope this can help.

Br. Frank

**Comment by Portisch** [comment]
@AlbertWeterings A meassurement on the original firmware would be helpfull. I know espurna will make repeats. But I don't know if the firmware of the EFM8 will do also repeats when transmitting. May I open another branch for testing repeating by the EFM8 chip.

**Comment by Portisch** [comment]
I have created a new branch. This version will repeat the data 5 times. It looks ok on the oscilloscope :
![scope_2](https://user-images.githubusercontent.com/3938570/37199430-e7825d04-2381-11e8-8ac1-096b1a589f64.png)
![scope_3](https://user-images.githubusercontent.com/3938570/37199441-eb0f6fa2-2381-11e8-8e2b-9124fcf3985e.png)

**Comment by AlbertWeterings** [comment]
@Portisch I will test it but will not be at home tomorrow morning.

As far as I can see ESPURNA in RAW supported mode:
Send message 1 time if only the message is comes in over the MQTT topic "rfraw"
Send message multiple times defined by {code},{times} over the MQTT topic "rfraw"

In regional mode:
Send messages from the web interface a predefined amount of times (default 4)
Send message 1 time if only the message is comes in over the MQTT topic "rfout" (not 100% sure)
Send message multiple times defined by {code},{times} over the MQTT topic "rfout"

So i'm not sure if it makes sense to have the EFM repeat the message as having it repeating it will not be able to receive during that time.

**Comment by Portisch** [comment]
@AlbertWeterings may you try the master again. I fixed the wrong byte position in the uart data.
Please test the master now. It doesn't include repeats - repeats are done by espurna.

**Comment by AlbertWeterings** [comment]
@Portisch Tested the latest version now and it still doesn't transmit anything. I also connected the input line from the receiver to decoding Arduino direct to Data line to transmitter and don't see anything passing by.

Did you program the light on pin 14 (P1.0) that it must blink while transmit? This stays off all the time.

I have send this several times to the EFM:
0D0A0D0AAAA5382C01CC057845D50C55
Response from EFM:
AAA055

In the morning I will hook up a oscilloscope to see if there is something coming out of the EFM

**Comment by AlbertWeterings** [comment]
@Portisch Last update and going to sleep now.

AAA50601D0F932113355
No RF light - Sometimes short beep - Nothing on receiver

AAA80601D0F932113355
No RF light - Nothing on receiver

AAB00601D0F932113355
RF light blinking - Nothing decodable receiver
Receiver gets data but can't decode

**Comment by AlbertWeterings** [comment]
@Portisch Well have been testing today and found all kind of strange behavior. I think we will have to sort out things more structural. I wrote a little program to send data into the EFM the same way ESPURNA does.
In every run I repeat the data 10 times. This is the result.
0D0A0D0AAA {A5} 382C01AE055045D50C550D0A
No LED blink - No buzzer - Data to transmitter - Receiver is not decoding

0D0A0D0AAA {A8} 382C01AE055045D50C550D0A
No LED blink - buzzer beeps for every transmission - No data to transmitter - Receiver is not decoding

0D0A0D0AAA {B0} 382C01AE055045D50C550D0A
No LED blink - buzzer beeps for every transmission - No data to transmitter - Receiver is not decoding

I think we shout first figure out how to get A5 mode send correct data that will be the original mode. Currently it sends something but unable to decode. A wild guess from me is lets try to invert the ASK signal.

**Comment by Portisch** [comment]
The receiver on the same RF Bridge get disabled while transmission. You only can see it on the oscilloscope. Or did you use another receiver?

**Comment by AlbertWeterings** [comment]
@Portisch I use a other receiver. And finally I figured out how to measure the pulses coming out of the EFM.
In A5 mode.
The signal of a PT2264:
Approximately : long pulses 1400us short pulses 430us (software measured) 
Same code from EFM measured the ASK line:
More precise : long pulses high 1300us low 1200us short pulses high 380us low 450us (oscilloscope) 

The signal of a EV1527:
Approximately : long pulses 1080us short pulses 327us (software measured) 
Same code from EFM measured the ASK line:
More precise : long pulses high 1000us low 840us short pulses high 280us low 350us (oscilloscope)

You can see there is a difference between the length of a pulse in low and high state,  I think that should not be the case.  Tomorrow I will measure the ASK signal on both transmitters of the remotes to get theire timings. I have also connected a different transmitter on the ASK line of the EFM that one did also not transmit a signal that could be decided.

**Comment by Portisch** [comment]
The length of the pulses should not matter. Only the duty cycle. 75% for a 1 bi, 25% for a  0 bit. Like 1400us and 430us are 23%.

**Comment by AlbertWeterings** [comment]
@Portisch I think I found where it goes wrong. I'm currently comparing the transmission from back to front.
When I send in mode A5 code 380E01D6058C45D50C it starts with a 7ms High 1.2ms low .56 ms high 14320 ms low and from there pulse length of 440 and 1440. The first part is wrong the remote is doing something completely different so I'm now comparing from the end to the front to see if all bits are transmitted.

**Comment by Portisch** [comment]
The 7ms high in front is needed to activate the SYN115 chip. The 1,2ms is low and than the sync pulse is starting. 14320 / 3968 *128 = 460us.

**Comment by AlbertWeterings** [comment]
@Portisch I don't have a oscilloscope that is possible to produce image. Can you have a look at the attached excel. It gives both the ASK out
[bridge-vs-remote.zip](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/1800401/bridge-vs-remote.zip)

I'm going to hook up a different transmitter to the bridge and see if that will transmit anything.

put on the remote and EFM

**Comment by ErikAndren** [comment]
Some questions based on http://www.princeton.com.tw/Portals/0/Product/PT2260_4.pdf
1. Looking at page 5 shouldn't the sync come last in the transaction? Maybe the command need to be sent twice?
2. Why divide by 3968? According to page 4 sync bit width is 4096 alpha

**Comment by AlbertWeterings** [comment]
it looks like sync bit comes in the end. what also is strange A0 is the first bit of the word and repeated just before the sync bit.

**Comment by ErikAndren** [comment]
Last A0 is most likely a typo. Check the examples just below the image (2 data, 4 data)

**Comment by Portisch** [comment]
@ErikAndren this is the unknown part. The master branch will do no repeats. The transmit_repeat branch will do 5 repeats on each UART command. I don't know how the original firmware is doing the transmission.

**Comment by AlbertWeterings** [comment]
Well I figured out the following see [Excel](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/1800492/bridge-vs-remote1.zip) file for details.
If I measure with a oscilloscope in a remote on the ASK line I get the same pattern as I receive with the RCSwitch library. **I only see a 0 on the end of the code Most likely the sync bit.** When I measure on the ASK line from the bridge I see the sync bit in front of the code and it is missing in the end.

**Comment by AlbertWeterings** [comment]
@Portisch & @ErikAndren guys the solution I disconnected the ASK line of the SYN115 from the EFM and injected the ASK signal as it comes from my PT2264 to the SYN115 and it transmit the code without any problem. So "The 7ms high in front is needed to activate the SYN115 chip" is not needed.

@Portisch Can you remove the 7ms high and place the SYNC puls on the end of the code. It must be working that way.

**Comment by Portisch** [comment]
So you will need a minimum of 2 repeats? Otherwise you don't know where the data is starting, isn't?

**Comment by AlbertWeterings** [comment]
Don’t think it is needed as the carrier of the SYN115 is always on so it will just start to transmit

**Comment by Portisch** [comment]
I don't think the SYN115 is always on. This is my problem. The datasheet is also telling something about power saving after ~75ms.

**Comment by AlbertWeterings** [comment]
@Portisch [Datasheet](http://www.electrodragon.com/w/images/b/b5/SYN113-SYN115-datasheet-version-1-1-.0.pdf) Page 6/16 and 8/16 "Enable control gates the ASK data. it only allows transmission when Lock, Amplitude and Under voltage detection conditions are valid. And page 10 14.2 first part

**Comment by AlbertWeterings** [comment]
75ms is standby delay this doesn’t say the SYN115 needs a puls to come out of standby it is saying the SYN115 goes into standby 75ms after the last high to low transition

**Comment by AlbertWeterings** [comment]
@Portisch you are right "So you will need a minimum of 2 repeats" Ï found in the) PT2264 data sheet that every code word has to be transmitted 4 times. This makes sense as the EV1527 has the sync pulse in front of the code word where PT2264 has it behind the code word.

In the transmission data to the EFM we put sync, high and low time. So the correct ASK signal can be generated as receivers for a specific type of transmitter will only listen to a signal with specific timings. 

EV1527 has a significant shorter sync pulse as PT2264 so I think you can use the sync pulse length to determine the type of transmitter on the receiver (decoding)  side. If you know the sync pulse you can use it to select the high and low times.

**Comment by Portisch** [comment]
@AlbertWeterings I moved the SYNC to the end an repeating the data 2 times right now. Please have a try: https://github.com/Portisch/RF-Bridge-EFM8BB1/tree/PT226x_Test

**Comment by mcspr** [comment]
@Portisch I am still waiting for external receivers, but tried out already known code for rf outlet associated with remote. New version with rotated sync works, but only when there are 4 repeats instead of 2. And that sometimes hangs the device - it stops handling uart, no receiving or sending commands via uart.

**Comment by Portisch** [comment]
@mcspr thx for this info. I will change the repeat to 4. I also have seen the hangup on the EFM8 device. I am right now searching how to fix this.

**Comment by AlbertWeterings** [comment]
@Portisch I will test the new version in the evening expect it will now work.

As far as I understand from the PT2264 datasheet the data needs to be send 4 times the Sync is in between the data so [data]sync[data]sync[data]sync[data]sync this is also what I see on the oscilloscope when I press the button on the remote in case of EV1527 I see the same on the oscilloscope only thing I did not notice is that for EV1527 sync is before data. In practice it doesn't matter if data is send 4 times with sync in between.

You say pulse length doesn't matter I'm not sure about that it doesn't matter if a specific receiver expects a predefined pulse length. On most uP the pulse width relates to the oscillator settings for example on a PT2264 you set the oscillator with a resistor on pin 15. On the EV1527 you do the same on pin 1. On the receiver side you set the oscillator to the same freq otherwise it will not receive right.So in basic you can't receive a code with a certain pulse width change the pulse width and expect the receiver to accept the code only based on duty cycle. 

I had a hard time catching the data from the bridge on my oscilloscope with the previous version as it was only send once and there was 500ms in between. Both remotes where easy as they transmit a constant stream.

**Comment by Portisch** [comment]
@mcspr Should work right now. I fixed the hangup and increased the repeat to 4 times.

**Comment by mcspr** [comment]
@Portisch Thanks. Will try out later.

I also remembered that espurna uses 4 times by default. That should probably be overriden to 1 there when RFRAW support is turned on. https://github.com/xoseperez/espurna/blob/8777d29f7c780bc14396a795c947d777b6d97262/code/espurna/config/general.h#L862

**Comment by AlbertWeterings** [comment]
@Portisch and @mcspr I know ESPURNA does the repeats but PT2264 documentation states no delay between transmission. ESPURNA has a default 500ms in between every transmission. So in different words I think best way to go is have EFM send repeat 4 times as in documentation of protocol and see this as 1 transmission (same way a remote does) , ESPURNA or other software can repeat the transmission with a delay (like pressing the remotes button several times)

**Comment by Portisch** [comment]
The new branch https://github.com/Portisch/RF-Bridge-EFM8BB1/tree/SYNC_IN_FRONT is doing the sync in front of data. Please test this and the https://github.com/Portisch/RF-Bridge-EFM8BB1/tree/PT226x_Test branch. But only with command 0xA5!

**Comment by AlbertWeterings** [comment]
@Portisch Thanks for the effort we are making progress.
The result.:
[https://github.com/Portisch/RF-Bridge-EFM8BB1/tree/SYNC_IN_FRONT](https://github.com/Portisch/RF-Bridge-EFM8BB1/tree/SYNC_IN_FRONT) Transmitting:
PT2264 code
380E01D6058C45D50C intermittently  2 out of 50 
382201D6056E45D50C intermittently  15 out of 50 (Changed timing)
382201D6056E45D503 correct 50 out of 50
EV1527 code 
2990012203D4B22894 correct 50 out of 50
2990012203D4B22898 correct 50 out of 50


[https://github.com/Portisch/RF-Bridge-EFM8BB1/tree/PT226x_Test](https://github.com/Portisch/RF-Bridge-EFM8BB1/tree/PT226x_Test) Transmitting:
PT2264 code
380E01D6058C45D50C correct
382201D6056E45D503 correct

EV1527 code 
2990012203D4B22894 correct
2990012203D4B22898 correct

So I'm a big fan already for the sync in front version. Only have to find why it not sends all codes correct will test with more after dinner.

**Comment by AlbertWeterings** [comment]
@Portisch Did a long test, it looks reasonably good.
The results:
EV1527:
Code learned recently 2968014A0424B22894 when I re-transmit I don't receive anything.
Code learned earlier    2990012203D4B22894 (timing is different) this transmits.
Received 50 out of 50 - 45 out of 50 correct.
Same code send with remote: 40 out of 50 correct.

Code learned recently 29720154041AB22898 when I re-transmit I don't receive anything.
Code learned earlier    2990012203D4B22898 (timing is different) this transmits.
Received 50 out of 50 - 40 out of 50 correct.
Same code send with remote: 45 out of 50 correct.

Code learned recently 2972015E041AB22891 when I re-transmit I don't receive anything.
Don't have a previous log on this code.
Same code send with remote: 45 out of 50 correct.

Code learned recently 298601540424B22892 when I re-transmit I don't receive anything.
Don't have a previous log on this code.
Same code send with remote: 44 out of 50 correct.

PT2264:
Code learned recently 382C01D6058245D5C0
Received 21 out of 50 - 17 out of 50 correct.
Same code send with remote: 47 out of 50 correct.

Code learned recently 385E01D6057845D5F0
Received 1 out of 50 - 1 out of 50 correct.
Same code send with remote: 42 out of 50 correct.

Code learned recently 389A01E0056E45D5CF
Received 50 out of 50 - 42 out of 50 correct.
Same code send with remote: 49 out of 50 correct.

Code learned recently 384001CC056E45D503
Received 50 out of 50 - 45 out of 50 correct.
Same code send with remote: 45 out of 50 correct.

Code learned recently 384001CC057845D50C
Received 50 out of 50 - 44 out of 50 correct.
Same code send with remote: 42 out of 50 correct.

Code learned recently 384001CC057845D530
Received 50 out of 50 - 42 out of 50 correct.
Same code send with remote: 44 out of 50 correct.

Code learned recently 385401D6057845D5C3
Received 50 out of 50 - 43 out of 50 correct.
Same code send with remote: 44 out of 50 correct.

Code learned recently 385401D6057845D5CC when I re-transmit I don't receive anything.
Don't have a previous log on this code.
Same code send with remote: 44 out of 50 correct.

Code learned recently 385E01D6057845D53C when I re-transmit I don't receive anything.
Code learned earlier    386801CC058245D53C (timing is different) this transmits.
Same code send with remote: 0 out of 50 correct.

Code learned recently 386801D6058245D533
Received 50 out of 50 - 45 out of 50 correct.
Same code send with remote: 40 out of 50 correct.

Code learned recently 387201D6057845D50F
Received 50 out of 50 - 41 out of 50 correct.
Same code send with remote: 46 out of 50 correct.

Code learned recently 388601E0057845D53F
Received 50 out of 50 - 42 out of 50 correct.
Same code send with remote: 42 out of 50 correct.

**Comment by Portisch** [comment]
Thx for testing. 

What does this mean:

> 298601540424B22892 when I re-transmit I don't receive anything.

Just learned and then transferred? Who is the receiver?

The sync_in_front is repeated 4 times. But because the sync have to be on the end, only 2 are useful.
First isn't complete because of DAY wakeup. Second and third are ok. Last one is missing the sync on the end. May more repeats are useful and also may a sent sync on end of the transfer would help.

**Comment by AlbertWeterings** [comment]
@Portisch . Answer to the first question is Learned with the bridge and transmitted by the bridge. My secondary receiver decoded the code during the learn process but as the bridge send it is not decoded.

I think with a sync on the of the "fourth code word" might help in case of the PT2264 but I don't know how it affects the EV1527. I would suggest to try if you make a version with this in it I can do the test over.

A other option might be Sync in front or on the end based on what we tell the EFM. Sync for PT2264 is different as for EV1527. I found there are 6 different protocols  with all their own sync pulse.

Maybe you found these two links already they give some inside information that might help improving the firmware.
[Home Automation – RF Protocols](http://tech.jolowe.se/home-automation-rf-protocols/)
[Home Automation – RF Protocol, update.](http://tech.jolowe.se/home-automation-rf-protocol-update/)

**Comment by Portisch** [comment]
@AlbertWeterings I think the problem would be the timer for the sync. Example code: 38 86 01 E0 05 78 45 D5 3F.

sync low time: 14470µs (0x3886)
sync high time: 467µs (0x3886 / (4096-128) * 128)
bit 0 high time: 480µs (0x1E0)
bit 0 low time: 1440µs (0x1E0 / 4 * 12)
bit 1 high time: 1400µs (0x578)
bit 1 low time: 467µs (0x578 / 12 * 4)
![scope_4](https://user-images.githubusercontent.com/3938570/37325947-333edd92-2690-11e8-9aeb-1f7ba8d510a3.png)
CH1 is the transmitted data, CH2 the received on the RF Bridge.
+Width(3) is 360µs - should be 470µs
-Width(3) is 14265µs - should be 14470µs

The bits should be in tolerance.
Bit 0:
![scope_6](https://user-images.githubusercontent.com/3938570/37326193-48fe8668-2691-11e8-8b4b-0def0b20ce8a.png)

Bit 1:
![scope_7](https://user-images.githubusercontent.com/3938570/37326205-546e17a2-2691-11e8-80d2-cee8a6939165.png)

**Comment by AlbertWeterings** [comment]
@Portisch I also just measured the Remotes:

PT2264 on 12v with 4.7M osc resistor.
 ___
|     |__________________________
Sync Thigh 560us Tlow 14400us Total sync time 14960us
The total bit time is 1880us short part of a bit 360us long part of a bit 1520us

EV1527 on 12v with 330K osc resistor
Sync Thigh 400us Tlow 10600us Total sync time 11000us
The total bit time is 1380us short part of a bit 350us long part of a bit 1030us

**Comment by AlbertWeterings** [comment]
Timings as decoded with @Portisch firmware.

EV1527

Tsync |   | Averange
-- | -- | --
Highest | 10760 | 10716.2
Lowest | 10620 |  
  |   |  
Tlow |   | Averange
Highest | 380 | 370.5
Lowest | 280 |  
  |   |  
Thigh |   | Averange
Highest | 1090 | 1082
Lowest | 990 |  

PT2264

Tsync |   | Averange
-- | -- | --
Highest | 14550 | 14505.9
Lowest | 14400 |  
  |   |  
Tlow |   | Averange
Highest | 500 | 487.8
Lowest | 430 |  
  |   |  
Thigh |   | Averange
Highest | 1440 | 1425.8
Lowest | 1340 |

**Comment by Portisch** [comment]
I pushed a new commit where I tuned the sync timer a little bit. My tests with UNDERWATER PAR56 LED LAMP are working without any problems. As the RF data will be repeated by the EFM8 6 times there is no need to repeat by espurna.
If the standard repeat of espurna is 4 the data will be sent 4 * 6 times.

**Comment by AlbertWeterings** [comment]
@Portisch I just did a short test and found that if I send a code 1 time it is decoded by the receiver twice if send 50 times it is decoded 100 times. So I think 6 repeats is a bit to much if a device has to switch on and off by the same code it will now do 2 actions.

Did you change the sync timer also on the receiving side? it looks to me during a short test the by the bridge decoded timing are more constant.

**Comment by AlbertWeterings** [comment]
@Portisch Sorry to say it became worse, most of the newly and old (by the bridge) learned codes are not decoded more than 4-9 times out of 50 after I transmit them again with the bridge.

I extended the test with a third receiver that has 12 channels and hooked up 12 led's to it. When I send the 12 codes corresponding to the led's (interval 1 sec) one after an other they shout light up one after an other in a steady flow. They don't it is not going regular and sometimes it skips led's.

An other observation is the RF led of the bridge (this is minor) it lights when receiving but doesn't light up when sending. In the original firmware id did. I think it is good to have it light also during transmitting.

**Comment by Portisch** [comment]
I merged the SYNC_IN_FRONT branch to the master. Changed repeat from 6 down to 4 and added the LED flashing while transmit.

**Comment by AlbertWeterings** [comment]
@Portisch Thanks will test in the evening. I also will get a few RF remotes today with hopefully different uPC's so we can check if they also work.

**Comment by AlbertWeterings** [comment]
@Portisch. Latest version in master branch is not working at all. I had to go some versions back to find one that is working a bit.

https://raw.githubusercontent.com/Portisch/RF-Bridge-EFM8BB1/master/Keil%208051%20v9.53%20-%20Release/RF_Bridge.hex
transmit (bad) Learn Very bad.
https://github.com/Portisch/RF-Bridge-EFM8BB1/blob/f9121860310382b6420c0c96509167578c2cca83/Keil%208051%20v9.53%20-%20Release/RF_Bridge.hex
transmit (no) Learn (no).
https://github.com/Portisch/RF-Bridge-EFM8BB1/commit/17fdda44ce018209819901c08382ff0c6fe7c0bf#diff-ff5aa20228c386165033a4574d0c48b6
transmit (no) Learn (no).

This one works reasonable:
https://github.com/Portisch/RF-Bridge-EFM8BB1/blob/6d73cdcd0849f78a4224ff0db69f5b91552b2bd8/Keil%208051%20v9.53%20-%20Release/RF_Bridge.hex
transmit (almost perfect) Learn (No)

So last one in the row sending almost perfect.
I got today a new bridge with original firmware. What I did is read all code's from the remote with this new bridge and programmed them in a loop with a 2 sec wait in between. I have the 12ch receiver programmed with the same codes in the same sequence. When I transmit the codes with the original firmware the led's switch on and of in the same sequence as the codes. When I send the codes with the firmware version that works reasonable the led's light up but sometimes one is skipped or they don't go in sequence.
 Learning codes is not working with this firmware the buzzer says beep but nothing appears on the screen.

Than more bad news. The EFM8 in the new bridge is protected only the first half of the flash rom can be read. A other problem Communication between ESP en EFM is not working with ESPURNA the board is slightly different designed. I will have to search what they did.

**Comment by Portisch** [comment]
You are sure it's protected or is it a 4k model? Would be bad because my firmware needs already ~6k. Do you have access to the RS232 header of the EFM8? Also a picture would be nice.

What I would need: transmit only one 0xA5 command over the RS232 to the EFM8 and record the T_DATA line with an oscilloscope. Then I can modify the firmware that it acts like the original one.

**Comment by AlbertWeterings** [comment]
I managed to get the original firmware from the EFM and wrote it back in a other one It took me some time but it works. The hex file is in a format you can't write it with the Arduino programmer it has to be manipulated as it is in the wrong format (lines are to long).

**Comment by AlbertWeterings** [comment]
Oke here it is the original firmware. The ZIP contains 3 files I tested them all and they work. You can write the INTELH8 file with the Arduino programmer the other files can be used for different programmers.

If someone wants to use a programmer like the Wellon VP-890 or EasyPro80B the EFM8 is not supported by the software select "Silicon Lab C8051F300 (ISP)" and your programmer will do the job over the ISP C2 bus. (do not program the config bit)

[SonOff_Bridge_origional_FW.zip](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/1813280/SonOff_Bridge_origional_FW.zip)

**Comment by Portisch** [comment]
Very good! Than I can start debugging tomorrow morning. Thank you! Is it also ok for you that I add it to the master branch?

**Comment by AlbertWeterings** [comment]
Sure these files I share so people can always decide to go back to the original firmware.

A remark on the beep function in the version I use it doesn't work yet. I have send in raw mode AAC001F455 to get a 500ms beep but nothing happens. The data between ESP and EFM has the linebreak in fron and after the code.

**Comment by Portisch** [comment]
It's just beeping, nothing more.

**Comment by Portisch** [comment]
Original firmware test command:
AA A5 38 86 01 E0 05 78 45 D5 3F 55

38 86:	14470µs
01 E0:	480µs
05 78:	1400µs

The original firmware is transmitting the data 8 times.
**The sync is in front of the data:**
![scope_12](https://user-images.githubusercontent.com/3938570/37452392-c1a8befe-2834-11e8-9f1b-5e26806697ec.png)
**Real timing of the sync bit:**
![scope_13](https://user-images.githubusercontent.com/3938570/37452397-c603540a-2834-11e8-8816-6ef19deb9303.png)
430µs ON, 12954µs OFF
**Calculated values:**
4096 Alpha = 13384µs
128 Alpha = 418µs
3968 Alpha = 12965µs

**Real timing of a bit 0:**
![scope_14](https://user-images.githubusercontent.com/3938570/37452406-d01f6140-2834-11e8-8e0c-5e9172df61df.png)
430µs ON, 1258µs OFF

**Calculated values:**
512 Alpha = 1688µs
128 Alpha = 422µs
384 Alpha = 1266µs

**Real timing of a bit 1:**
![scope_15](https://user-images.githubusercontent.com/3938570/37452476-fc7c78b8-2834-11e8-8da5-cc3e2cda0187.png)
1254µs ON, 435µs OFF

**Calculated values:**
512 Alpha = 1689µs
128 Alpha = 422µs
384 Alpha = 1266µs

**_Alternative firmware:_**
**Real timing of the sync bit:**
![scope_16](https://user-images.githubusercontent.com/3938570/37452483-02962762-2835-11e8-956d-e2ada567a858.png)
476µs ON, 14431µs OFF

**Calculated values:**
4096 Alpha = 14907µs
128 Alpha = 465µs
3968 Alpha = 14441µs

**Real timing of a bit 0:**
![scope_17](https://user-images.githubusercontent.com/3938570/37452486-0830265a-2835-11e8-801e-b4239b3c9f1a.png)
471µs ON, 1445µs OFF

**Calculated values:**
512 Alpha = 1916µs
128 Alpha = 479µs
384 Alpha = 1437µs

**Real timing of a bit 1:**
![scope_18](https://user-images.githubusercontent.com/3938570/37452494-0cbf7144-2835-11e8-99f1-832bebeba8fa.png)
1391µs ON, 473µs OFF

**Calculated values:**
512 Alpha = 1864µs
128 Alpha = 466µs
384 Alpha = 1398µs

So now the question is:
The sync should be 14470µs. On the original firmware it get transmitted with these values:
430µs ON, 12954µs OFF
4096 Alpha = 13384µs

This is a difference about 1100µs...

**Comment by Portisch** [comment]
There is also a difference at the sniffing. It looks like the original firmware is filtering repeated data.
My firmware isn't.
When I send a code repeated 8 times every second this is the result:

Original firmware:
> 12:17:08.693: AAA43EA801FE060EAAAAAA55
> 12:17:09.590: AAA43EA801FE060EAAAAAA55
> 12:17:10.220: AAA43EC601FE060EAAAAAA55
> 12:17:11.747: AAA43EA801FE060EAAAAAA55

My firmware:
> 12:19:48.487: AAA43EA801CC05E6AAAAAA55
> 12:19:48.619: AAA43EA801F4060EAAAAAA55
> 12:19:48.685: AAA43E9E01FE0618AAAAAA55
> 12:19:48.751: AAA43E9E0208060EAAAAAA55
> 12:19:48.817: AAA43E9E0208060EAAAAAA55
> 12:19:49.068: AAA43E9E01F4060EAAAAAA55
> 
> 12:19:50.014: AAA43E9E01E005FAAAAAAA55
> 12:19:50.081: AAA43EBC01EA05FAAAAAAA55
> 12:19:50.147: AAA43EB201FE0604AAAAAA55
> 12:19:50.213: AAA43E9E02080604AAAAAA55
> 12:19:50.279: AAA43E9E01FE060EAAAAAA55
> 12:19:50.345: AAA43E9401F40604AAAAAA55
> 12:19:50.592: AAA43E9E0208060EAAAAAA55

**Comment by AlbertWeterings** [comment]
@Portisch I will test the rebuild this evening.

I also noticed during read/learn the original firmware needs more samples, it looks like it records a few samples and only send one that occurred the most to the UART

This morning I did not had much time but i increased the baud rate from my receiver to the PC and noticed that I got like 350 received codes when I send only 50 that is what you see in the data and most likely is the repeating the EFM8 does.

Also the RF led behaves different in the original firmware. It blinks 1 time short after send in alternative firmware it does during send data. Changing this might be safer as no CPU cycles will be lost for controlling the LED during send. In receive mode it looks like the LED blinks when a sync comes by.

In my Setup beeping did not work yesterday I will check with newer version.

Maybe it is good to add a command to ask the EFM8 witch version of the alternative FW is loaded this can become handy when working on reported issues.

**Comment by AlbertWeterings** [comment]
@Portisch Found a bug in the latest build.
After sending the beep command "AAC001F455" The firmware doesn't come back in listening mode so receiver is dead. It comes back to life after sending a "A5" command.

Learning a code isn't working very well:
It doesn't learn any of the EV1527 remote codes and only some after many tries with the PT2264

**Comment by Portisch** [comment]
Give me the two codes that I can test it.

**Comment by AlbertWeterings** [comment]
@Portisch The difference you see in timings can they be related to the internal oscillator I have seen the EFM8 has 3 internal oscillators. The higest freq is also the most accurate 1.5% vs 2% For sampling the receiver data the highest possible frequency will be the most accurate.

**Comment by Portisch** [comment]
Better would be an oscilloscope record of the not working remotes. May it is timing/tolerance issue.

**Comment by AlbertWeterings** [comment]
PT2264
387C01CC058245D503 Need several tries
38A401E0058245D533 Does not learn at all

EV1527
2990015E0410B22894 Does not learn at all
2990015E0410B22898 Does not learn at all
2990015E0406B22891 Does not learn at all
299A015E0410B22892 Does not learn at all

All codes read by original firmware

**Comment by AlbertWeterings** [comment]
Something else the Bridge with original firmware is not learning the codes I transmit with the other switch using alternative firmware. I will now try to feed the transmitting bridge the codes read with the original bridge and see what happens.

**Comment by Portisch** [comment]
I will need oscilloscope records of the R_DATA line. Then and only than I can see the real duty cycle and high/low times.

**Comment by AlbertWeterings** [comment]
not from ask line in remote? I think that will be the most accurate

**Comment by AlbertWeterings** [comment]
The original FW bridge has it difficult learning its own codes when transmitted with the alternative firmware

**Comment by AlbertWeterings** [comment]
Be right back, I will measure both in receiver and in remote.

**Comment by Portisch** [comment]
> not from ask line in remote? I think that will be the most accurate
Better would be the rdata line. That's what the EFM8 see. It doesn't matter what firmware, only the waveform would be interesting.

**Comment by AlbertWeterings** [comment]
I’m working on it

**Comment by AlbertWeterings** [comment]
@Portisch 

PT2264 |   |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | --
  | Sync |   | Byte | Byte | 1 | Byte | 0
  | H | L |   | Low | High | Low | High
Remote | 490 | 14560 | 1900 | 1440 | 460 | 480 | 1400
RF | 520 | 14480 | 1900 | 1490 | 410 | 530 | 1360


EV1527 |   |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | --
  |   | Sync | Byte | Byte | 1 | Byte | 0
  | H | L |   | Low | High | Low | High
Remote | 320 | 1076 | 1400 | 1060 | 370 | 360 | 1040
RF | 360 | 1072 | 1390 | 1110 | 300 | 400 | 980


Screenshots will follow. All times are uS

**Comment by Portisch** [comment]
@AlbertWeterings Thank you for the nice pictures ;)
I increased the tolerance of the sync bit detection.
Before it was:

sync |  
-- | --
high | low
520 | 14480

sync high time|  
-- | --
min | max
420 | 514

Now with commit aba0b61258238b57a9c1c1b86edf486fbf8905d4 it is:

sync high time|  
-- | --
min | max
374 | 561

I don't know why the EV1527 doesn't get recognized because on your sample the sync should have already been fitted:

sync |  
-- | --
high | low
360 | 10720

sync high time|  
-- | --
min | max
311 | 380

Also the bit 0 and bit 1 tolerances should be good.
On your PT226x example:

high | low
-- | --
78,4% | 28%

On your EV1527 example:

high | low
-- | --
79,8% | 29%

I do have a 8% tolerance from 75% for bit 1 and 25% for bit 0.

**Comment by AlbertWeterings** [comment]
@Portisch : I will do a new test this evening. Did you also looked at the other issues I mentioned above
https://github.com/Portisch/RF-Bridge-EFM8BB1/issues/2#issuecomment-373420655
and
https://github.com/Portisch/RF-Bridge-EFM8BB1/issues/2#issuecomment-373461815

**Comment by Portisch** [comment]
The LED flashing takes no or less CPU cycles. It changes on the state of the received bit.
Same on sending. Won't fix.

The other thing with restarting sniffing after command 0xC0 and 0xFF is fixed.

**Comment by AlbertWeterings** [comment]
@Portisch Good news and a bit less good.

ESPURNA 1.12.4  (Raw support on)
If learn key in web interface is pressed mode A1 EV1527, PT2264 and LD27L2 codes appear in the web interface.

ESPURNA 1.12.4  (Raw support off)
Send 1 for on 0 for off to topic ESPURNA_xxxx/rflearn/{relay_id}/set
EV1527, PT2264 and LD27L2 codes appear in the web interface. (Learn mode A1)

ESPURNA 1.12.4  (Raw support on)
Send 1 for on 0 for off to topic ESPURNA_xxxx/rflearn/{relay_id}/set
EV1527, PT2264 and LD27L2 codes appear in the web interface. (Learn mode A1)

ESPURNA 1.12.4  (Raw support on)
Send AAA655 to topic ESPURNA_xxxx/rfraw/set to enables Raw sniffing (Mode A5).
- EV1527, LD27L2 and PT2264 codes received are send in topic ESPURNA_xxxx/rfin to MQTT broker
Send AAA755 to topic ESPURNA_xxxx/rfraw/set to disable Raw sniffing (returns to default sniffing)
- EV1527, LD27L2 and PT2264 codes received are send in topic ESPURNA_xxxx/rfin to MQTT broker

ESPURNA 1.12.4  (Raw support on)
Send AAA955 to topic ESPURNA_xxxx/rfraw/set to enables LEARN NEW (Mode A9).
EFM goes into mode A9 When code 385401B8054645D503 is broadcasted EFM response with AA AB 04 01 45 D5 03 55 ESPURNA doesn't do anything with it. (only visual in Telnet when debug line in void _rfbReceive of rfbridge.ino is enabled)

ESPURNA 1.12.4  (Raw support on)
Send AAC001F455 to topic ESPURNA_xxxx/rfraw/set to enable BUZZER (Mode C0).
Buzzer beeps LONGER as the expected 01 F4 = 500ms. EFM returns into default sniffing mode. ESPURNA does not understand to repeat beeps if ,x is added (like it can repeat transmitting codes)

ESPURNA 1.12.4  (Raw support on)
Send AAFF55 to topic ESPURNA_xxxx/rfraw/set to request alternative FW version (Mode FF).
EFM8 returns the FW version in HEX (only visual in Telnet when debug line in void _rfbReceive of rfbridge.ino is enabled)

**Comment by AlbertWeterings** [comment]
The less good news.
The codes from EV1527:
297C0168042EB22894
298601680424B22898
297C01680424B22891
297C01680424B22892
Are not transmitted in a correct way, my secondary receiver can decode them when send with the remote but not when send with the EFM in Raw and default mode after learning them.

**Comment by Portisch** [comment]
Thank you for testing! Looks like PT226x is working. For the EV1527 I will need the original waveform. Best from the remote. These screens aren't complete above. The timing of Bit 0 and 1 is missing.

**Comment by AlbertWeterings** [comment]
@Portisch Well I have put some time today in finding a working solution based on the data sheets of PT2264 and EV1527 remotes. To compare practical working and theory I used specific receivers for PT2264 and one for EV1527. And started to generate signals according to data sheets instead of signals transmitted by remotes and came to some nice conclusions.

For PT2264 The data sheet say's:
Sync High part = 1/8 bit time
Sync Low part = 4 * bit time
Bit time = 4 * period time
This means all times can be calculated out of 1 variable the "Period time" so If we keep to the rule of how the total signal has to be build there is a min and max period time on what the receiver is working. I tested this for hours and found min = 383 and max = 1024 uS with repeat to 4 the signal was rock solid on all values in between. At 703 uS the repeats could go back to 2 without loosing stability.

For EV1527 The data sheet say's:
Sync High part = period time
Sync Low part = 31 * period time
Bit time = 4 * period time
This means all times can be calculated out of 1 variable the "Period time" so If we keep to the rule of how the total signal has to be build there is a min and max period time on what the receiver is working. I tested this for hours and found min = 316 and max = 441 uS with repeat to 8 the signal was rock solid on all values in between. At 378 uS the repeats could go back to 3 without loosing stability.

The period time is what we know as Thigh coming from the EFM when it receives the value has to be divided by 3 to get 1 period time.

Here is the Arduino Sketch I used to test on a Arduino Leonardo with a standard low-cost transmitter:
[Emulated_Remote.zip](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/1821965/Emulated_Remote.zip)

**Comment by ciotlosm** [comment]
@franki29 are you using the web interface of Espurna? Did you enable "raw" flag when flashing?

**Comment by franki29** [comment]
Hi, no i am using the standard bin version of Espurna, i did not compile by myself this time.
And yes, i am using the web interface to switch ON/OFF.
It looks like the new PCB has a additional 5 Pin chip, a"IB3HJ" and it is close to the SYN115. But i could not find what kind of chip it is in the internet yet.

**Comment by ciotlosm** [comment]
For me the new code for some buttons I have at work doesn't work. It doesn't show up after pressing learn. Some DIO buttons, model: CAWST-907.

**Comment by franki29** [comment]
Some more information:
Learn code with old PCB ON
2710014A03D4051511
New PCB 
2738013603CA051511

OFF
Old PCB 26E8014003D4051514
New PCB 2738014003C0051514

But even i use the OLD PCB code in the new PCB it does not work with the new PCB.

**Comment by AlbertWeterings** [comment]
@Portisch 
With the values in the table below I think you can set the receive limits of both remotes. All values are calculated with the values I found when testing witch actual RF receivers. (See above)

PT2264 | Period Min | Period Max | Sync High Min | Sync High Max | Sync Low Min | Sync Low Max | Bit 1 High Min | Bit 1 High Max | Bit 1 Low Min | Bit 1 Low Max | Bit Time Min | Bit Time Max | Repeats | Transmit time Min | Transmit time Min
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  |   |   |   |   |   |   | Bit 0 Low Min | Bit 0 Low Max | Bit 0 High Min | Bit 0 High Max |   |   | by EFM | ms | ms
  | 395 | 1057 | 197.5 | 528.5 | 6122.5 | 16383.5 | 1185 | 3171 | 395 | 3171 | 1580 | 4228 | 8 | 353.92 | 947.072
  |   |   |   |   |   |   |   |   |   |   |   |   |   |   |  
EV1527 | Period Min | Period Max | Sync High Min | Sync High Max | Sync Low Min | Sync Low Max | Bit 1 High Min | Bit 1 High Max | Bit 1 Low Min | Bit 1 Low Max | Bit Time Min | Bit Time Max | Repeats | Transmit time Min | Transmit time Min
  |   |   |   |   |   |   | Bit 0 Low Min | Bit 0 Low Max | Bit 0 High Min | Bit 0 High Max |   |   | by EFM | ms | ms
  | 325 | 444 | 325 | 444 | 9750 | 13320 | 975 | 1332 | 325 | 1332 | 1300 | 1776 | 8 | 330.2 | 451.104


Now we have to find a way to transmit the signals in the correct format based on the hex data we send to the EFM. I found that if the period time (Thigh/3) is higher than 444us it is alway's a PT226x based remote below 444us it can be a EV1527 and PT2264. To get to know below 444us what remote format to use is harder I think I found a way but can't confirm yet, it looks like the decimal "code" value of a EV1527 has always 1 digit more than the PT2264. Will try to find confirmation in the datasheets.

It is also a possibility to send the codes in both formats PT226x and EV1527 but that can lead to having devices respond to codes that supposed to go to a different device.

**Comment by ErikAndren** [comment]
Very nice. 
Maybe it makes more sense to have a byte in the uart command to the EFM define what protocol is used instead of a heuristics based approach? What will happen if yet another format with a similar timing is to be used?

**Comment by AlbertWeterings** [comment]
@ErikAndren I like the idea. Problem is the original FW can do the trick without sending a extra Byte so to stay backwards compatible the most general protocols shout work without sending a extra byte.

In RAW mode I think having the extra Byte would be a good thing to make sure the data is send in the proper shape.

**Comment by Jason2866** [comment]
@franki29  My new unknow 5 Pin chip is namend "IB3HB"

![rfb](https://user-images.githubusercontent.com/24528715/37594338-2ccf124e-2b75-11e8-8fec-8bc289ca1f90.jpg)

**Comment by Portisch** [comment]
@AlbertWeterings 

> For PT2264 The data sheet say's:
> Sync High part = 1/8 bit time
> Sync Low part = 4 * bit time - Sync High Part
> Bit time = 4 * period time

This is Correct:
Let's say Tsync = 10000µs. Then Tsync_high = Tsync / (4096 - 128) * 128 = 322µs.
Tsync_high = 10000µs.
The bit timing is 3:1 or 1:3.

> For EV1527 The data sheet say's:
> Sync High part = period time
> Sync Low part = 30 * period time
> Bit time = 4 * period time

Not correct! Sync = 32 * period time. It's hard to read but the arrows in the datasheet shows 1/32 and 31/32 for the sync.
Let's say Tsync = 10000µs. Then Tsync_high = Tsync / 31 = 322µs.
Tsync_high = 10000µs.
The bit timing is 3:1 or 1:3.

So I see no difference between except the sync. For PT226x it should be on back, for EV1527 it is in the front.

@Jason2866
This looks like a USB ESD protection. The SYN115 is connected to the same pin like here: https://github.com/Portisch/RF-Bridge-EFM8BB1/blob/master/doc/Bridge_v1_0.jpg
Also the SYN470 pin is the same.

**Comment by Jason2866** [comment]
@Portisch ok, thx. So new and old version should in general work?!
What frequence does X2 on your board have? On the new board it is 13.560 Mhz.
In circuit diagramm it is 12.000 Mhz.

**Comment by Jason2866** [comment]
@Portisch: You are right it is a step up controller. Pin 3 on SYN115 is 12 Volt.

**Comment by ErikAndren** [comment]
I don't follow this math:
_Let's say Tsync = 10000µs. Then Tsync_low = Tsync / (4096 - 128) * 128 = 322µs._

According to datasheet tsync is 4096 alpha, ts_high is 128 alpha: 
Tsync_high = (tsync / 4096) * 128 = (10000 / 4096) * 128 = 312.6 us.
Tsync_low = (tsync / 4096) * (4096 - 128) = 9687.5 us

**Comment by Portisch** [comment]
@ErikAndren 
I (we) guess the Tsync in the command 0xA5 is the low time of the sync.
The sync bit is 128&#945; high + 3968&#945; low == 4096&#945;

I checked the EV1527 command 297C0168042EB22894:
T-Sync-low: 10620µs
T-Bit-0-high: 360µs
T-Bit-1-high: 1070µs

So T-Sync-high should be 10620 / (4096 - 128) * 128 == 342µs
T-Bit-0-low: 360 * 3 == 1080µs
T-Bit-1-low: 1070 / 3 == 356µs

Sync:
![scope_0](https://user-images.githubusercontent.com/3938570/37596524-b7c5ef4c-2b7c-11e8-82ac-ab087a257cbe.png)

Bit 1:
![scope_1](https://user-images.githubusercontent.com/3938570/37596532-bf32ebea-2b7c-11e8-84a5-cd8a39a19c76.png)

Bit 0:
![scope_2](https://user-images.githubusercontent.com/3938570/37596535-c4a2d838-2b7c-11e8-942a-800dfb3fae83.png)

**Comment by ErikAndren** [comment]
ok, thank you.
Did you come to this conclusion by measuring with the original firmware?

**Comment by Portisch** [comment]
I have done now some more tests with the original firmware:


AA A5 1F 40 01 90   04 B0 B2 28 94 55 |   |   |  
-- | -- | -- | --
  | Should | Meassured |  
  |   | High | Low
  | [µs] | [µs] | [µs]
Tsync | 8000 | 360 | 7164
Tlow | 400 | 360 | 1079
Thigh | 1200 | 1075 | 363
  |   |   |  
  |   |   |  
AA A5 2E E0 02 58 07 08 B2 28 94 55 |   |   |  
  | Should | Meassured |  
  |   | High | Low
  | [µs] | [µs] | [µs]
Tsync | 12000 | 537 | 10744
Tlow | 600 | 538 | 1616
Thigh | 1800 | 1612 | 542

data | description
-- | --
AA | sync init
-- | --
A5 | command
1F 40 | Tsync
01 90 | Tlow
04 B0 | Thigh
B2 28 94 | data
55 | sync end



Make the sync:
High: Tlow;
Low: Tsync

Bit 0:
High: Tlow;
Low: Thigh

Bit 1:
High: Thigh
Low: Tlow

**Comment by ErikAndren** [comment]
Interesting.
What is the end goal with the 0xA5 command from your perspective? 
Is it to replicate the original firmware implementation or to make it follow the protocol specifications as close as possible?

**Comment by Portisch** [comment]
The master is doing like the datasheet is telling us. But it looks like not working 100%.
So if we get the 0xA5 command like the original firmware it should be OK. Maybe after this it can be improved again. But for now it is less working than the original firmware.

The alternative firmware have to be 100% compatible to the original one to be able to work also with an espurna firmware what isn't supporting the RAW data.

**Comment by Jason2866** [comment]
@ErikAndren Good question. Protocol Specs. are made for best compatibility. But real life is sometimes different. Maybe the guys at ITEAD made an error or the copied (wrong) code "everybody" use...
In my humble opinion the code from the origi. Firmware is "fast written" -> it works! Specs who cares...
So in my case a very old and simple code (Intertechno Power Outlet) is false decoded from the RFBridge
and cannont learnend. To sent works very well, with hand calculated correct code.
This brings me to the conclusion, the code isnt that good...

**Comment by AlbertWeterings** [comment]
@Portisch I agree on the part where for EV1527 Tsync low is 31* period time.
Can we also agree that what the EFM reports as Thigh = 1/3 of a bit(other words the high part of a bit 1)? If we do than perid time is 1/4 of a bit. If you start calculating from these facts the total Sync of a PT226x is completely different as from the EV1527.

If I generate the signals with a arduino they are perfectly accepted by the specific receivers to the EV1527 and PT226x I tried to put the sync as calculated for the EV1527 behind the code part of the PT226x and the receiver doesn't respond anymore.

PT226x |   |   |   |   |   | EV1527 |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Periode time | TSync | Tsync High | Tsync low | Thigh from EFM |   | Periode time | Total Sync | T Sync High | T Sync low | Thigh from EFM
  |   | 1/8*bit | (4*bit)-(1/8*bit) |   |   | Thigh/3 | Thigh*32 | Thigh | Thigh *31 |  
395 | 6320 | 197.5 | 6122.5 | 1185 |   | 395 | 12640 | 395 | 12245 | 1185
396 | 6336 | 198 | 6138 | 1188 |   | 396 | 12672 | 396 | 12276 | 1188
397 | 6352 | 198.5 | 6153.5 | 1191 |   | 397 | 12704 | 397 | 12307 | 1191
398 | 6368 | 199 | 6169 | 1194 |   | 398 | 12736 | 398 | 12338 | 1194
399 | 6384 | 199.5 | 6184.5 | 1197 |   | 399 | 12768 | 399 | 12369 | 1197
400 | 6400 | 200 | 6200 | 1200 |   | 400 | 12800 | 400 | 12400 | 1200
401 | 6416 | 200.5 | 6215.5 | 1203 |   | 401 | 12832 | 401 | 12431 | 1203
402 | 6432 | 201 | 6231 | 1206 |   | 402 | 12864 | 402 | 12462 | 1206
403 | 6448 | 201.5 | 6246.5 | 1209 |   | 403 | 12896 | 403 | 12493 | 1209
404 | 6464 | 202 | 6262 | 1212 |   | 404 | 12928 | 404 | 12524 | 1212
405 | 6480 | 202.5 | 6277.5 | 1215 |   | 405 | 12960 | 405 | 12555 | 1215
406 | 6496 | 203 | 6293 | 1218 |   | 406 | 12992 | 406 | 12586 | 1218
407 | 6512 | 203.5 | 6308.5 | 1221 |   | 407 | 13024 | 407 | 12617 | 1221
408 | 6528 | 204 | 6324 | 1224 |   | 408 | 13056 | 408 | 12648 | 1224
409 | 6544 | 204.5 | 6339.5 | 1227 |   | 409 | 13088 | 409 | 12679 | 1227
410 | 6560 | 205 | 6355 | 1230 |   | 410 | 13120 | 410 | 12710 | 1230
411 | 6576 | 205.5 | 6370.5 | 1233 |   | 411 | 13152 | 411 | 12741 | 1233
412 | 6592 | 206 | 6386 | 1236 |   | 412 | 13184 | 412 | 12772 | 1236
413 | 6608 | 206.5 | 6401.5 | 1239 |   | 413 | 13216 | 413 | 12803 | 1239
414 | 6624 | 207 | 6417 | 1242 |   | 414 | 13248 | 414 | 12834 | 1242
415 | 6640 | 207.5 | 6432.5 | 1245 |   | 415 | 13280 | 415 | 12865 | 1245
416 | 6656 | 208 | 6448 | 1248 |   | 416 | 13312 | 416 | 12896 | 1248
417 | 6672 | 208.5 | 6463.5 | 1251 |   | 417 | 13344 | 417 | 12927 | 1251
418 | 6688 | 209 | 6479 | 1254 |   | 418 | 13376 | 418 | 12958 | 1254
419 | 6704 | 209.5 | 6494.5 | 1257 |   | 419 | 13408 | 419 | 12989 | 1257
420 | 6720 | 210 | 6510 | 1260 |   | 420 | 13440 | 420 | 13020 | 1260
421 | 6736 | 210.5 | 6525.5 | 1263 |   | 421 | 13472 | 421 | 13051 | 1263
422 | 6752 | 211 | 6541 | 1266 |   | 422 | 13504 | 422 | 13082 | 1266
423 | 6768 | 211.5 | 6556.5 | 1269 |   | 423 | 13536 | 423 | 13113 | 1269
424 | 6784 | 212 | 6572 | 1272 |   | 424 | 13568 | 424 | 13144 | 1272
425 | 6800 | 212.5 | 6587.5 | 1275 |   | 425 | 13600 | 425 | 13175 | 1275
426 | 6816 | 213 | 6603 | 1278 |   | 426 | 13632 | 426 | 13206 | 1278
427 | 6832 | 213.5 | 6618.5 | 1281 |   | 427 | 13664 | 427 | 13237 | 1281
428 | 6848 | 214 | 6634 | 1284 |   | 428 | 13696 | 428 | 13268 | 1284
429 | 6864 | 214.5 | 6649.5 | 1287 |   | 429 | 13728 | 429 | 13299 | 1287
430 | 6880 | 215 | 6665 | 1290 |   | 430 | 13760 | 430 | 13330 | 1290
431 | 6896 | 215.5 | 6680.5 | 1293 |   | 431 | 13792 | 431 | 13361 | 1293
432 | 6912 | 216 | 6696 | 1296 |   | 432 | 13824 | 432 | 13392 | 1296
433 | 6928 | 216.5 | 6711.5 | 1299 |   | 433 | 13856 | 433 | 13423 | 1299
434 | 6944 | 217 | 6727 | 1302 |   | 434 | 13888 | 434 | 13454 | 1302
435 | 6960 | 217.5 | 6742.5 | 1305 |   | 435 | 13920 | 435 | 13485 | 1305
436 | 6976 | 218 | 6758 | 1308 |   | 436 | 13952 | 436 | 13516 | 1308
437 | 6992 | 218.5 | 6773.5 | 1311 |   | 437 | 13984 | 437 | 13547 | 1311
438 | 7008 | 219 | 6789 | 1314 |   | 438 | 14016 | 438 | 13578 | 1314
439 | 7024 | 219.5 | 6804.5 | 1317 |   | 439 | 14048 | 439 | 13609 | 1317
440 | 7040 | 220 | 6820 | 1320 |   | 440 | 14080 | 440 | 13640 | 1320
441 | 7056 | 220.5 | 6835.5 | 1323 |   | 441 | 14112 | 441 | 13671 | 1323
442 | 7072 | 221 | 6851 | 1326 |   | 442 | 14144 | 442 | 13702 | 1326
443 | 7088 | 221.5 | 6866.5 | 1329 |   | 443 | 14176 | 443 | 13733 | 1329
444 | 7104 | 222 | 6882 | 1332 |   | 444 | 14208 | 444 | 13764 | 1332

**Comment by Portisch** [comment]
@AlbertWeterings 
Did you try the new branch? This should now act like the original firmware on comand 0xA5.

**Comment by Portisch** [comment]
As the transmit like described in the datasheet isn't working very well I have done this new PT226x branch. I think ITEAD have done trial & error on the firmware. So the alternative firmware should "simulate" as near as possible the original behavior.

But I don't know if this will also have affect to the transmitting:

  | should | is | diff | diff
-- | -- | -- | -- | --
  | [µs] | [µs] | [µs] | [%]
Tsync | 8000 | 7164 | 836 | 10,45
Tlow | 400 | 360 | 40 | 10
Thigh | 1200 | 1075 | 125 | 10,41667
  |   |   |   |  
  | should | is | diff | diff
  | [µs] | [µs] | [µs] | [%]
Tsync | 12000 | 10744 | 1256 | 10,46667
Tlow | 600 | 538 | 62 | 10,33333
Thigh | 1800 | 1612 | 188 | 10,44444

The original firmware reduce the timings about 10%.

This is with the PT226x branch:

  | should | is | diff | diff
-- | -- | -- | -- | --
  | [µs] | [µs] | [µs] | [%]
Tsync | 8000 | 7987 | 13 | 0,1625
Tlow | 400 | 392 | 8 | 2
Thigh | 1200 | 1188 | 12 | 1

**Comment by AlbertWeterings** [comment]
@Portisch The PT2264 branch is better than the master It learns PT2264 remote (pretty well sometimes need retries) and instantly learns EV1527. It transmits PT2264 reliable (receiver responds by only the repeats from the EFM) EV1527 reliable (receiver responds by only the repeats from the EFM)

Codes from SonOff PIR-2 are learned ok ( Chip LD27L2 )
Codes from SonOff door sensor are learned ok ( EV1527 ) 

ESPURNA may not repeat otherwise receiver is triggered twice. So codes cannot be used in the webinterface.

This looks really good thanks for the effort. Can you tell how the values TSync, Tlow and Thigh are now related to the generated signal?

**Comment by Portisch** [comment]
Just take a look above: https://github.com/Portisch/RF-Bridge-EFM8BB1/issues/2#issuecomment-374219015

**Comment by Portisch** [comment]
@All
Thank all for the help!
Issue closed by commit 3ee74778bc6501768c847abb222eba6c6ebd13b4.

**Comment by Jason2866** [comment]
@Portisch 
Is master branch now the same as 3ee7477 ?

**Comment by Portisch** [comment]
Yes, the commit got merged to the master.
